scripts/pre_push_hook.py

```diff
@@ -0,0 +1,184 @@
+    """Get all files that are modified and are being pushed"""
+    # get branch name from e.g. local_ref='refs/heads/lint_hook'
+    branches = [ref.local_ref.split("/")[-1] for ref in ref_list]
+    hashes = [ref.local_sha1 for ref in ref_list]
+    remote_hashes = [ref.remote_sha1 for ref in ref_list]
+    changed_files = set()
+    for branch, sha1, remote_sha1 in zip(branches, hashes, remote_hashes):
+        # if remote doesn't have the branch yet, git reports the following
+        #  remote_sha1: '0000000000000000000000000000000000000000'
+        if set(remote_sha1) != {"0"}:
+            try:
+                files = _compare_to_remote(REMOTE, branch)
+            except ValueError as e:
+                sys.exit(e.message)
+        else:
+            # Get the difference to origin/develop instead
+            try:
+                files = _compare_to_remote(REMOTE, branch,
+                                           remote_branch="develop")
+            except ValueError:
+                # give up, return all files in repo
+                try:
+                    files = _git_diff_names_only(GIT_NULL_COMMIT, sha1,
+                                                 diff_filter="ACMRT")
+                except ValueError as e:
+                    sys.exit(e.message)
+
+        changed_files.update(files)
+
+    return list(changed_files)
+
+
+def _get_refs():
+    # Git provides refs in STDIN
+    ref_list = [GitRef(*ref_str.split()) for ref_str in sys.stdin]
+    print "ref_list:"
+    pprint.pprint(ref_list)
+    return ref_list
+
+
+def _start_linter(files):
+    script = os.path.join(SCRIPTS_DIR, LINTER_SCRIPT)
+    task = subprocess.Popen([PYTHON_CMD, script, LINTER_FILE_FLAG] + files)
+    task.communicate()
+    return task.returncode
+
+
+def _start_sh_script(scriptname):
+    cmd = ["bash", os.path.join(SCRIPTS_DIR, scriptname)]
+    task = subprocess.Popen(cmd)
+    task.communicate()
+    return task.returncode
+
+
+def _install_hook():
+    # install script ensures that oppia is root
+    oppia_dir = os.getcwd()
+    hooks_dir = os.path.join(oppia_dir, ".git", "hooks")
+    pre_push_file = os.path.join(hooks_dir, "pre-push")
+    if os.path.islink(pre_push_file):
+        print "Symlink already exists"
+        return
+    try:
+        os.symlink(os.path.abspath(__file__), pre_push_file)
+        print "Created symlink in .git/hooks directory"
+    # raises AttributeError on windows, OSError added as failsafe
+    except (OSError, AttributeError):
+        shutil.copy(__file__, pre_push_file)
+        print "Copied file to .git/hooks directory"
+
+
+def main():
+    if _ARGS.install:
+        _install_hook()
+        sys.exit(0)
+    refs = _get_refs()
+    modified_files = _collect_files_being_pushed(refs)
+    print "Files being pushed:"
+    pprint.pprint(modified_files)
+    lint_status = _start_linter(modified_files)
+    if lint_status != 0:
+        sys.exit("Push failed, please correct the linting issues above")
```

Comment:
Might be better if we do a print of the message, then `sys.exit`. @seanlip, second opinion please.